### PR TITLE
Implement new GeoInterface traits

### PR DESCRIPTION
**Pull request by evetion**
[GeoInterface 1.0](https://juliageo.org/GeoInterface.jl/stable/) makes use of Traits, this is a PR to update our pre-1.0 GeoInterface implementation.

Note that I had to add some methods to align with GeoInterface:
- Added `getm` method, previously not wrapped
- Added `getgeom` for LineStrings. Technically OGR doesn't expose this (and wants us to use getpoint), but our `ngeom` does have a switch for Lines.
- Added `is3d` and `ismeasured` to be used when getting the the third or fourth coordinate
- Added `keys` and `values` to a Feature

I've also added dependencies on `GeoInterfaceRecipes`, a plot package based on `GeoInterface`, which previously was integrated into `GeoInterface` and on `Extents`, a package defining bounding boxes as `NamedTuples`.

I also note that much in ArchGDAL doesn't support Z en M coordinates properly, for which I will create a separate issue. In short, many of our creation methods don't result in the correct type (just a Point instead of Point25D etc.). This makes it hard to correctly implement `getcoord` for the `m` dimension, as it could be both the third or fourth dimension, depending on the type.

### CODE REVIEW

**Comment by evetion** [comment]
- [ ] Benchmark CI doesn't seem to update to latest GeoInterface (still using `v0.5.7`?)
- [ ] Plots recipes are required for the documentation to work @rafaqz

**Comment by visr** [comment]
> Benchmark CI doesn't seem to update to latest GeoInterface (still using v0.5.7?)

That'd be because of Shapefile.jl

https://github.com/yeesian/ArchGDAL.jl/blob/e401e0079d97dac9d4755a4bc9dce345a9c5bf81/benchmark/Project.toml#L6

File: src/geointerface.jl
Comment (visr) [inline src/geointerface.jl:250]: Base.convert?

**Comment by evetion** [comment]
@rafaqz 

Can you take a look here at the `convert` options? I keep running into ambiguous things, as our suggetion for implementing `convert` overrides *everything*, including existing methods such as those for GeoFormatTypes.

```julia
  Expression: sprint(print, convert(AG.IGeometry, json)) == "Geometry: POINT (100 70)"
  MethodError: convert(::Type{ArchGDAL.IGeometry}, ::GeoFormatTypes.GeoJSON{String}) is ambiguous. Candidates:
    convert(::Type{T}, source::X) where {T<:ArchGDAL.AbstractGeometry, X<:GeoFormatTypes.GeoJSON} in ArchGDAL at /Users/evetion/code/ArchGDAL.jl/src/convert.jl:48
    convert(::Type{T}, geom::X) where {T<:ArchGDAL.IGeometry, X} in ArchGDAL at /Users/evetion/code/ArchGDAL.jl/src/geointerface.jl:262
  Possible fix, define
    convert(::Type{T}, ::X) where {T<:ArchGDAL.IGeometry, X<:GeoFormatTypes.GeoJSON}
```

**Comment by rafaqz** [comment]
Ww need to dispatch both on `ArchGDAL.AbstractGeometry`. That will work.

**Comment by evetion** [comment]
> We need to dispatch both on ArchGDAL.AbstractGeometry. That will work. Both first arguments need to be equally specific.

Well that's what I did first, but it segfaults preparedgeometry tests(!). Something to investigate, but otherwise, I'm leaning towards the idea that only IGeometries should be created from conversion, while _all_ geometries can function as a source. But maybe better as an issue than to keep changing more and more in this PR.

**Comment by rafaqz** [comment]
Both can be on `IGeometry` as well I guess? I cant remember the distinctions honestly.

File: src/geointerface.jl
Comment (yeesian) [inline src/geointerface.jl:127]: Out of curiosity, when is it Z, versus M?

**Comment by yeesian** [comment]
> I'm leaning towards the idea that only IGeometries should be created from conversion, while all geometries can function as a source

Yeah I think this might be a viable approach (or at least it was how I was mentally thinking about things without properly having articulated them anywhere)

File: src/geointerface.jl
Comment (evetion) [reply src/geointerface.jl:127]: Well I made a mistake here, I think we should call `getz` here. It could be `m`, but only when `is3d` is false and `ismeasured` true?

**Comment by evetion** [comment]
Benchmark should be able to run when https://github.com/JuliaGeo/Shapefile.jl/pull/72 is merged and released.

File: src/geointerface.jl
Comment (yeesian) [inline src/geointerface.jl:151]: Can we add unit tests for the operations introduced from this line onwards? Other than that, I think it's fine that some of the new traits in `GeoInterface.geomtrait(...)` aren't being covered in this PR.

**Review by yeesian — APPROVED** [review]
Overall LGTM, thank you so much!

Might be good to update the top level comment of this PR with the changes.

File: src/geointerface.jl
Comment (yeesian) [reply src/geointerface.jl]: Addressed by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/96f9731f18daec06b1d9c07a98903917aa8fa8ce

File: src/geointerface.jl
Comment (rafaqz) [inline src/geointerface.jl:3]: This could be `const`

File: src/geointerface.jl
Comment (rafaqz) [inline src/geointerface.jl:59]: This is a little confusing, maybe refactor with named variables?

**Review by rafaqz — COMMENTED** [review]
Few minor comments, otherwise LGTM

File: src/geointerface.jl
Comment (visr) [reply src/geointerface.jl:3]: `const Dict{DataType, Function}`

**Comment by evetion** [comment]
Ok, I can't merge without review approval here. The benchmark suite seems to require a released ArchGDAL version to test against.

**Review by yeesian — APPROVED** [review]
LGTM modulo the remaining open comments, thank you so much for seeing this through!

File: src/geointerface.jl
Comment (yeesian) [reply src/geointerface.jl]: Resolved by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/3079fd31974a6ef5ccbf4241ee4b5bc2c4bfca2b

File: src/geointerface.jl
Comment (yeesian) [reply src/geointerface.jl]: Resolved by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/3079fd31974a6ef5ccbf4241ee4b5bc2c4bfca2b

File: src/geointerface.jl
Comment (yeesian) [reply src/geointerface.jl:145]: Resolved by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/e2892d0d5a8d9a725c0a051ec173237f49e1638e

File: src/geointerface.jl
Comment (yeesian) [reply src/geointerface.jl]: Resolved by https://github.com/yeesian/ArchGDAL.jl/pull/290/commits/a56959cddc9fbd0cf5a90f1f158e07fd7f37d7f3

**Pull request merged**
